src/lib/cert/x509/key_constraint.cpp

```diff
@@ -54,11 +25,11 @@ Key_Constraints find_constraints(const Public_Key& pub_key,
       constraints |= KEY_AGREEMENT;
 
    if(name == "RSA" || name == "ElGamal")
-      constraints |= KEY_ENCIPHERMENT | DATA_ENCIPHERMENT;
+      constraints |= KEY_ENCIPHERMENT | DATA_ENCIPHERMENT | ENCIPHER_ONLY | DECIPHER_ONLY;
 
    if(name == "RSA" || name == "RW" || name == "NR" ||
-      name == "DSA" || name == "ECDSA")
-      constraints |= DIGITAL_SIGNATURE | NON_REPUDIATION;
+      name == "DSA" || name == "ECDSA" || name == "ECGDSA" || name == "ECKCDSA")
+      constraints |= DIGITAL_SIGNATURE | NON_REPUDIATION | KEY_CERT_SIGN | CRL_SIGN;
```

Comment:
Change applied in 6cbff45. I took the time to extract some typical key usage scenarios from RFC 5280 and add unit tests for `verify_cert_constraints_valid_for_key_type()`. During which I corrected that ENCIPHER_ONLY and DECIPHER_ONLY were allowed for RSA and ElGamal, where RFC 5280 says that both are undefined if not used together with KEY_AGREEMENT. This change was only introduced in this PR though, so no worries.

`$ ./botan gen_self_signed rsa.priv localhost` now creates a certificate with no key usage extension, which is what we wanted.